lib/krane/global_deploy_task.rb

```diff
@@ -0,0 +1,207 @@
+      Krane::StatsD.distribution('all_resources.duration', Krane::StatsD.duration(start),
+        tags: statsd_tags << "status:timeout")
+      raise
+    rescue Krane::FatalDeploymentError => error
+      logger.summary.add_action(error.message) if error.message != error.class.to_s
+      logger.print_summary(:failure)
+      Krane::StatsD.event("Deployment failed",
+        "One or more resources failed to deploy to #{context}",
+        alert_type: "error", tags: statsd_tags + %w(status:failed))
+      Krane::StatsD.distribution('all_resources.duration', Krane::StatsD.duration(start),
+        tags: statsd_tags << "status:failed")
+      raise
+    end
+
+    private
+
+    def deploy!(resources, verify_result, prune)
+      prune_whitelist = []
+      resource_deployer = ResourceDeployer.new(task_config: @task_config,
+        prune_whitelist: prune_whitelist, max_watch_seconds: @global_timeout,
+        selector: @selector, statsd_tags: statsd_tags)
+      resource_deployer.deploy!(resources, verify_result, prune)
+    end
+
+    def validate_configuration
+      task_config_validator = GlobalDeployTaskConfigValidator.new(@task_config,
+        kubectl, kubeclient_builder)
+      errors = []
+      errors += task_config_validator.errors
+      errors += @template_sets.validate
+      errors << "Selector is required" unless @selector.present?
+      unless errors.empty?
+        add_para_from_list(logger: logger, action: "Configuration invalid", enum: errors)
+        raise TaskConfigurationError
+      end
+
+      logger.info("Using resource selector #{@selector}")
+      logger.info("All required parameters and files are present")
+    end
+    measure_method(:validate_configuration)
+
+    def validate_resources(resources)
+      validate_globals(resources)
+
+      Concurrency.split_across_threads(resources) do |r|
+        r.validate_definition(@kubectl, selector: @selector)
+      end
+
+      resources.select(&:has_warnings?).each do |resource|
+        record_warnings(logger: logger, warning: resource.validation_warning_msg,
+          filename: File.basename(resource.file_path))
+      end
+
+      failed_resources = resources.select(&:validation_failed?)
+      if failed_resources.present?
+        failed_resources.each do |r|
+          content = File.read(r.file_path) if File.file?(r.file_path) && !r.sensitive_template_content?
+          record_invalid_template(logger: logger, err: r.validation_error_msg,
+            filename: File.basename(r.file_path), content: content)
+        end
+        raise FatalDeploymentError, "Template validation failed"
+      end
+    end
+    measure_method(:validate_resources)
+
+    def validate_globals(resources)
+      return unless (namespaced = resources.reject(&:global?).presence)
+      namespaced_names = namespaced.map do |resource|
+        "#{resource.name} (#{resource.type}) in #{File.basename(resource.file_path)}"
```

Comment:
This has come up a few times so I've added a test `test_global_deploy_validation_catches_namespaced_cr`. But resource.global? works because [KubernetesResource::build]( https://github.com/Shopify/kubernetes-deploy/pull/602/files#diff-58f1540757c3323310e495f480799ebdL56) set's the resources `global` ivar based on the list of all global kinds fetched from the cluster ("api-resources", "--namespaced=false"). So there isn't actually a need to parse it out of the CRD.